onset_maproom/calc.py

```diff
@@ -53,6 +332,228 @@ def water_balance(
+    reduced (or actual) evapotranspiration,
+    drain
+    """
+    # Start water balance ds with runoff
+    if runoff is None:
+        runoff = xr.zeros_like(daily_rain).rename("runoff")
+        runoff.attrs = dict(description="Runoff", units="mm")
+    water_balance = xr.Dataset().merge(runoff)
+    # Compute Effective Precipitation
+    peffective = (daily_rain - water_balance.runoff).rename("peffective")
+    peffective.attrs = dict(description="Effective Precipitation", units="mm")
+    water_balance = water_balance.merge(peffective)
+    # Additional vars and coords
+    et = (et * xr.ones_like(water_balance.peffective)).rename("et")
+    water_balance = water_balance.merge(et)
+    if kc_params is not None:
+        for adim in kc_params.dims:
+            for thedims in water_balance.dims:
+                if adim != thedims and adim != "kc_periods":
+                    water_balance[adim] = kc_params[adim]
+    if p_d is not None:
+        for adim in p_d.dims:
+            for thedims in water_balance.dims:
+                if adim != thedims and adim != time_coord:
+                    water_balance[adim] = p_d[adim]
+    # Get time_coord info
+    time_coord_size = water_balance.peffective[time_coord].size
+    # Intializing sm, et_crop and et_crop_red
+    # Creating variables and broadcasting everybody
+    soil_moisture = xr.full_like(
+        water_balance.peffective, np.nan
+    ).rename("soil_moisture")
+    soil_moisture.attrs = dict(description="Soil Moisture", units="mm")
+    water_balance = water_balance.merge(soil_moisture)
+    et_crop = xr.full_like(
+        water_balance.peffective, np.nan
+    ).rename("et_crop")
+    et_crop.attrs = dict(description="Crop Evapotranspiration", units="mm")
+    water_balance = water_balance.merge(et_crop)
+    (water_balance,) = xr.broadcast(water_balance)
+    water_balance["soil_moisture"] = water_balance.soil_moisture.copy()
+    water_balance["et_crop"] = water_balance.et_crop.copy()
+    # Give time dimension to sminit
+    sminit0 = xr.full_like(
+        water_balance.soil_moisture
+        .isel({time_coord: 0}).expand_dims(dim=time_coord),
+        sminit,
+    )
+    sminit0 = sminit0.assign_coords(
+        {time_coord: sminit0[time_coord] - np.timedelta64(1, "D")}
+    )
+    # Initialize Ks
+    ks = 1
+    if rho is not None:
+        raw = calibrate_available_water(taw, rho)
+        ks = single_stress_coeff(
+            sminit0,
+            taw,
+            raw.isel({time_coord: 0}, missing_dims="ignore"),
+            time_coord=time_coord,
+        ).squeeze(time_coord)
+    # Create or Initialize Kc
+    if kc_params is None:
+        kc0 = xr.ones_like(water_balance.et)
+    else:
+        if p_d is not None:
+            kc0 = kc_interpolation(p_d, kc_params, time_coord=time_coord)
+        else:
+            p_d_find = planting_date(
+                sminit0, sm_threshold, time_coord=time_coord
+            ).expand_dims(dim=time_coord)
```

Comment:
In the PR description you said,

> Planting Date is of same nature as onset/cessation_date which is that it has a T grid that from which the variable expressed in timedeltas refers to. Right now there is only one planting date (see more below in After this PR section). I realized when I added planting date variable (p_d) to the water_balance dataset, that I had to change its dim name not to be T, otherwise it would be meaninglessly extended over the other variables T. So that makes me think that there is maybe a flaw in the mental conception on how those _date variables should be... to be discussed

But here you're using `time_coord` as the dimension name. Has this problem been resolved?

I don't actually understand why the planting date array has a time dimension at all. It seems like it should only vary by X and Y.

Reply:
The reference point of the timedelta could be stored as an attribute rather than a dimension.

Reply:
Yes so I think that one last sentence suggests you understood what I meant. in onset/cessation, we used a dimension to store the reference time point. And it is typically the same day of the year, year after year. So here I used the same model because at some point (next step after this PR), we'll need a planting date that varies with years too. Planting date can vary with years, obviously if it's estimated within the lo; but also if the user gives it as input. For instance the user could use onset_date to estimate planting dates and feed that as input. Or they could have independent planting date data from somewhere else.

Reply:
oh yes i see... if years is a dimension from splitting, then the reference day-month can be a constant attribute. As things are now, the 2 are combined on a coord.

Splitting into 2 dimensions will need a bit more subtlety than managing the leap years because it will also need to cover the case where a "year" is not a "calendar year" because the season of interest overlaps 2 "calendar years." I guess that is why the grouping was appealing because it doesn't require to have a workable 2D array for years/doy; you just need the mechanism to "tag" the days you care about.

But I am game with either way.

I can clarify the T name business. At the very end, I merge p_d in the dataset to be part of the outputs, because its dimension is called T, the resulting p_d is expanded over the whole daily T of the other variables (with I guess many NaNs if I recall). That is why I renamed T of p_d just before merging.

So yeah.. it's funny that the very first things we had to start coding are the ones that are the most subtle with managing time dimensions...

Right now, I don't have a strong feeling for which route to take. If there is going to be reduction, groupby seems more straightforward, if not, moving to 2D seems more relevant, with all the subtleties mentioned (and maybe more).

Reply:
In the FBF application the season months (e.g. DJF) are a constant, and we have a year dimension. We define the year to be the year of the midpoint of the season, so T=2000 refers to the season that runs Dec 1999 to Feb 2000 (whose midpoint is Jan 15 2000). (I think it would have been more convenient if we had used the start or end of the season instead of the midpoint.) Seems like we could do the same here.

We can try the 2D approach first for this case (in a subsequent PR, not this one), where groupby isn't applicable, and then we'll see how we like it and decide whether we want to rewrite onset date to use that approach instead of groupby.

Thanks for explaning the dimension name business, I understand now.